lib/TimeSeries/AdaptiveFilter.pm

```diff
@@ -0,0 +1,210 @@
+
+use List::Util qw(sum max);
+
+use parent qw/Exporter/;
+
+our @EXPORT_OK = qw/filter/;
+
+=head1 SYNOPSYS
+
+  use TimeSeries::AdaptiveFilter qw/filter/;
+
+  # creation with defaults
+  my $filter = filter();
+
+  # create filter with tuned parameters
+  my $filter = filter({
+    floor             => 6
+    cap               => 0.2,
+    lookback_capacity => 20,
+    lookback_period   => 4,
+    decay_speeds      => [0.03, 0.01, 0.003],
+    build_up_count    => 5,
+    reject_criterium  => 4,
+  });
+
+  # usage
+  my $now = time;
+  $filter->($now, 100.002);        # returns true, i.e. all data is valid on learning period
+  $filter->($now + 1, 100.001);    # returns true
+  ...                              # it learns form sample of 60 seconds
+  $filter->($now + 60, 100.005);   # returns true
+  $filter->($now + 61, 99.9995);   # returns true, as value does not differs much
+  $filter->($now + 62, 10_0000);   # returns false, outlier data
+  $filter->($now + 63, 10.0001);   # returns false, outlier data
+  $filter->($now + 64, 100.011);   # returns true, even if the sample is oulier, because
+                                   # the filter rejected too much values, and has to
+                                   # re-adapt to time seria again
+
+=head1 DESCRIPTION
+
+For the details of underlying mathematical model of the filter, configurable paramters
+and their usage, please, look at the shipped C<doc> folder.
+
+=cut
+
+
+# private lexical function, which calculates MAD (mean absolute deviation)
+my $_mad = sub {
+    my ($mad, $ad, $trust, $density, $ds) = @_;
+    my $mu = $trust * (1 - exp(-$density * $ds));
+    return (1 - $mu) * $mad + $mu * $ad;
+};
+
+my $sqrt2pi = 1 / sqrt(2 * atan2(1, 1));
+
+sub filter {
+    my $params = shift;
+
+    ##############################
+    ## filter tuning parameters ##
+    ##############################
+
+    # minimum amount of values in lookback to take an decision.
+    # Otherwise, input values will be accepted
+    my $floor = $params->{floor} // 6;
+    # maximum share of rejected input values. Upon hit, input values will be accepted
+    my $cap = $params->{cap} // 0.2;
+    # maximum amount ot input values in lookback
+    my $lookback_capacity = $params->{lookback_capacity} // 20;
+    # the retention period for lookback, i.e. max age for input values
+    my $lookback_period  = $params->{lookback_period}  // 4;
+    my $decay_speeds     = $params->{decay_speeds}     // [0.03, 0.01, 0.003];
+    my $build_up_count   = $params->{build_up_count}   // 5;
+    my $reject_criterium = $params->{reject_criterium} // 4;
+
+    ########################
+    ## enclosed variables ##
+    ########################
+
+    my $lookback          = [];
+    my $lookback_rejected = 0;
+    my $minute            = [];
+    my $trust_w;
+    my $ad;
+    my $ads = [];    # used on build stage only
+    my $mad;
+    my $mads;
+    my $wsum = 0;
+    my $csum = 0;
+    my $vol;
+    my $_accepted;
```

Comment:
why are `lookback`, `minute` and `ads` array references? I guess they could equally well be normal lists. That would save an indirection when accessing the values.